File: solution/502. IPO/502.cpp
class Solution {
public:
    int findMaximizedCapital(int k, int w, vector<int>& profits, vector<int>& capital) {
        int n = profits.size();
        vector<pair<int, int>> A(n);
        for (int i = 0; i < n; ++i) {
            A[i] = {capital[i], profits[i]};
        }
        sort(A.begin(), A.end());
        
        priority_queue<int> pq;
        int i = 0;
        while (k--) {
            while (i < n && w >= A[i].first) {
                pq.push(A[i].second);
                i++;
            }
            if (pq.empty()) break;
            w += pq.top(); pq.pop();
        }
        return w;
    }
};

File: solution/502. IPO/502.py
class Solution:
    def findMaximizedCapital(self, k: int, w: int, profits: List[int], capital: List[int]) -> int:
        heap = []
        A = sorted(zip(capital, profits), key=lambda p: p[0])
        i = 0
        for _ in range(k):
            while i < len(A) and A[i][0] <= w:
                heapq.heappush(heap, -A[i][1])
                i += 1
            if not heap:
                break
            w -= heapq.heappop(heap)
        return w
    
